### PR TITLE
disabled add liquidity when min and max are out of range

### PR DESCRIPTION
**Pull request by p6te**
Disabled 'add liquidity' when MIN and MAX are out of pool range.  Show the grey slider in a range that is disabled. 
image

### CODE REVIEW

**Review by wojciech-cichocki — CHANGES_REQUESTED** [review]
It will be better to change priority of these errors message. The incorrect concentration level should definitely be addressed first, as someone initially arranges the position setup, and then adjusts the token proportions afterward.

image

**Review by wojciech-cichocki — CHANGES_REQUESTED** [review]
I've rechecked many cases, and there is one scenario we should handle.

Assuming we have an incorrect concentration level and we switch the liquidity provision method from "concentration" to "range", this incorrect range gets transferred

File: src/components/NewPosition/RangeSelector/RangeSelector.tsx [hunk 1]
```diff
@@ -74,7 +82,7 @@ export const RangeSelector: React.FC<IRangeSelector> = ({
```
Comment (wojciech-cichocki) [inline src/components/NewPosition/RangeSelector/RangeSelector.tsx:82]:
```Typescript
  const [leftRange, setLeftRange] = useState(MIN_TICK)
  const [rightRange, setRightRange] = useState(MAX_TICK)
  ```

Overall, `MAX_TICK` and `MIN_TICK` shouldn't be used at all. Instead, use `getMaxTick(tickSpacing)` and `getMinTick(tickSpacing)`

**Review by wojciech-cichocki — CHANGES_REQUESTED** [review]

File: src/components/NewPosition/NewPosition.tsx [hunk 2]
```diff
@@ -349,6 +414,11 @@ export const NewPosition: React.FC<INewPosition> = ({
     }
   }
 
+  const concentrationArray = useMemo(
```
[the remark this answers is not in the document]
Comment (p6te) [reply src/components/NewPosition/NewPosition.tsx:417]: no, i have moved it from the child component here

File: src/components/NewPosition/NewPosition.tsx [hunk 3]
```diff
@@ -216,15 +223,32 @@ export const NewPosition: React.FC<INewPosition> = ({
   }
 
   const onChangeRange = (left: number, right: number) => {
-    setLeftRange(left)
-    setRightRange(right)
+    const leftMax = isXtoY ? getMinTick(tickSpacing) : getMaxTick(tickSpacing)
```
[the remark this answers is not in the document]
Comment (p6te) [reply src/components/NewPosition/NewPosition.tsx:226]: I have created getMarkersInsideRange function and used it in changeRangeHandler and onChangeRange, getMinSliderIndex doesn't work exactly the same

File: src/components/NewPosition/DepositSelector/DepositSelector.tsx [hunk 4]
```diff
@@ -57,6 +57,10 @@ export interface IDepositSelector {
   priceALoading?: boolean
   priceBLoading?: boolean
   feeTierIndex: number
+  concentrationArray: number[]
+  concentrationIndex: number
+  minimumSliderIndex: number
+  isConcentrated: boolean
```
[the remark this answers is not in the document]
Comment (p6te) [reply src/components/NewPosition/DepositSelector/DepositSelector.tsx]: applied

File: src/components/NewPosition/DepositSelector/DepositSelector.tsx [hunk 5]
```diff
@@ -143,6 +151,12 @@ export const DepositSelector: React.FC<IDepositSelector> = ({
       return 'Insufficient lamports'
     }
 
+    if (concentrationIndex < minimumSliderIndex && isConcentrated) {
+      return concentrationArray[minimumSliderIndex]
```
[the remark this answers is not in the document]
Comment (p6te) [reply src/components/NewPosition/DepositSelector/DepositSelector.tsx]: applied

File: src/components/NewPosition/RangeSelector/RangeSelector.tsx (hunk 1 above)
Comment (p6te) [reply src/components/NewPosition/RangeSelector/RangeSelector.tsx:89]: applied

File: src/components/NewPosition/DepositSelector/DepositSelector.tsx [hunk 6]
```diff
@@ -171,7 +185,15 @@ export const DepositSelector: React.FC<IDepositSelector> = ({
...
     return 'Add Liquidity'
-  }, [tokenAIndex, tokenBIndex, tokenAInputState.value, tokenBInputState.value, tokens])
+  }, [
+    tokenAIndex,
+    tokenBIndex,
+    tokenAInputState.value,
+    tokenBInputState.value,
+    tokens,
+    isConcentrated,
+    concentrationIndex
```
[the remark this answers is not in the document]
Comment (p6te) [reply src/components/NewPosition/DepositSelector/DepositSelector.tsx]: yes, you're right

File: src/components/NewPosition/NewPosition.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (wojciech-cichocki) [reply src/components/NewPosition/NewPosition.tsx:226]: How are the differences there?

[the remark this answers is not in the document]
Comment (p6te) [reply src/components/NewPosition/NewPosition.tsx:226]: I have refactored the function to get ticks inside the range and reused it in getMinSliderIndex as well.

**Review by wojciech-cichocki — APPROVED** [review]

**Pull request merged**